### ISSUE TITLE
Digital Badging Integration

### ISSUE THREAD

**Issue by joshdanielson** [loconomics/loconomics#746]
## Epic Story 
As Loconomics, we want to issue digital badges to students and display any external badges they have so they can attract more clients.

**Comment by ndarilek** [comment]
The components on this branch now have the following behaviors:
 * Previously, you had to manually append ".json" to all URLs attached to badges. Now, the plain URL returned by Badgr is sufficient. Components now correctly set the `Accept:` header to `application/json`.
 * A consequence is that you now need to check the "2.0" version when generating share URLs. Would be nice if Badgr defaulted this somehow, and we may wish to ask them about that.
 * Unfortunately, Badgr collections don't seem to link to V2.0 assertions by default. For now I hack around this and perform an API request for each badge. Badgr should probably let you specify collection versions for consistency with assertions. I'll ask them about this. Once they've implemented this, we can switch back to the quicker method of displaying the badge data returned by a single API call rather than making one per badge.
 * I tweaked the template for badges a bit. The badge image now links to the badge itself on Badgr. This may be a more compact alternative to displaying all badge details in our own interface. No strong feelings about which approach is best, I just implemented this to give us some choices.

**Comment by joshdanielson** [comment]
@IagoSRL I'm going to make a general task list for this issue that we can discuss tomorrow:
- Review [documentation (work in progress)](https://github.com/loconomics/loconomics/issues/770). We will issue badges for licenses, background checks, verifications, and performance so that it's in the badge format no matter what the credential. Users will request validation of their licenses the same but the output will be a badge.
- [ ] create UserBadge database table using this [Google Doc](https://docs.google.com/spreadsheets/d/1D1AnVWKXOG2w6WkeGPTSL0-51Xn_NugbrPlcpvqu_P8/edit?usp=sharing)
- [ ] create API call for /badge-add activity (badge-editor component) with create, edit, delete
- [ ] add to API call (not sure which one this is: https://dev.loconomics.com/api/v1/en-US/users/540?includeFullJobTitleID=-1&_=1518478033436) to include in jobProfile a section entitled "credentials" that fetches the info in UserBadge from badgr.io. See @ndarilek's components (badge-collection and badge-view). In doing so, if the entry in UserBadge is of Type 'collection' use the code in collection to retrieve each individual badge url so that in 'credentials' is an array of individual badges. Include in the query both the selected jobTitleID and "-1". We can talk through this tomorrow.
- [ ] Help decide on edit/delete logic. If loconomics issues the badge (ModifiedBy != 'user') then the edit icon should not show and the user won't be able to delete or edit. If they added it, they can.
- [ ] Review new activities to ensure their functionality. Here's a quick summary:
- listingEditor: badges are shown as Credentials with a link to view full details (badge-view) and edit if the user created it (badge-edit). The "Add" button for Credentials leads to a new activity: 'credentials-add'
- credentials-add: is where users can add professional licenses (link to licenseCertifications as before with same upload process) or add Digital badges (badge-add). We'll at some point put background checks and verifications (think email, name, identification, address, etc.) here.
- listing: badges are shown under Credentials
- badge-view: full details of the badge
- [ ] Review new badge components (there should only be two in my mind. I left @ndarilek 's work that should be incorporated when fetching data for listing and listingEditor
 - editor: allows ONLY user-generated digital badges to be edited. 
 - viewer: views only single badges...collections should already have been fetched and their individual badges included in the object.
 - collection and view: @ndarilek 's work that connects directly with badgr.io

**Comment by joshdanielson** [comment]
@IagoSRL this may also help you by looking at the data of a finished badge:
https://badgr.io/public/assertions/XmE3ncMeRP-1JLr7LfGd_w
https://api.badgr.io/public/assertions/XmE3ncMeRP-1JLr7LfGd_w.json?v=2_0

**Comment by joshdanielson** [comment]
@IagoSRL I had some doubts on the UserBadge table and ultimately think we should go with the following schema
`CREATE TABLE dbo.UserBadge
(UserBadgeID int NOT NULL,
UserID int NOT NULL,
SolutionID int NOT NULL,
BadgeURL varchar(2078) NOT NULL,
Type varchar(25) NOT NULL,
Category varchar(25) NOT NULL,
ExpiryDate datetime,
CreatedDate datetime NOT NULL,
ModifiedDate datetime NOT NULL,
CreatedBy nvarchar(12) NOT NULL,
Active bit DEFAULT 1 NOT NULL,
PRIMARY KEY (UserBadgeID))`

I chose SolutionID instead of UserListingID because licenses and background checks will need to be checked by solution, not a listing, if that makes sense. We will need a general SolutionID (thinking it be named "General") to map things that apply to many solutions OR can just be null. Then when we pull credentials for a listing, we will based on the solutions in the listing AND null and group them by solution. Users can also add badges/collections by solution (similar to what we want offerings to be). 

Let me know your thoughts or anything that needs clarification before I make any updates to the front end.

**Comment by IagoSRL** [comment]
Can make sense to do it by solution, since solutions seems to be the new cornerstone, enabling setup and capabilities (rather than job title as until now). My doubts, similar to offerings, is about if repetitions can happens (same badges making sense to be related to several solutions, or being required by several solutions); if that happens, we could still create requirements of badges (licenses, checks,..) based on solutions added by the user to any listing, but still let them be linked to a listing (where it should be displayed), the trick here is the table where we set-up which badges we require at which solutions, and have unique ID for the required-badge (required licenses-credentials and background checks before), that should be included in the UserBadge table.

*Just some thoughts before going to sleep, so take them carefully*

**Comment by joshdanielson** [comment]
@IagoSRL to scale it will be nearly impossible to map out the specific badge(s) required so I think relying on a manual check is fine for now and maybe adding an "Approved" flag to UserListingSolution (wondering if this should really be UserSolution but then gets confusing)? I was almost thinking we could map the approved badgeID in that table but it could be more than one and then would require a new table and not be much use. So I think there needs to be a manual requirement for solutions with either CredentialCheckRequired or BackgroundCheckRequired flagged that staff checks and approves the badges and it's done. If neither is flagged, the "Approved" is automatically set to 1.

Then when we pull the info for Credentials in the listing, we can somehow de-dupe and organize by solution with one extra category "General" or something like that. 

Maybe that all that gets us back to the same problem.

**Comment by IagoSRL** [comment]
The "Approval" process conflicts with allowing users to provide a badge's collection. We commented that a collection is mutable (can change it's content at any point), that means we cannot really approve it because may get new badges that were not approved and may even get unapproved when reviewed. We can make it an 'optimistic' process, where we approve the collection (by approving all the badges inside at a specific point in time) and expect the approval to be valid until next time we reviewed again for new content (a recurrent task, checking whether there is new content manually); could be done automatically (detect new badges in the collection), but on that case we need to record the badges inside the collection anyway --and is the only realistic way to do it rather than manually to let us scale. That leads to the same idea I was having: we need to record each badge inside a collection rather than the collection itself, and approve them individually.
In an initial process, we can communicate to the user to re-submit a collection when it adds new badges to it, so can appear at their listing.
In a further step, we can set a recurrent task to check collections (server side) and see if there is new content, registering them and proceed same way as if user had uploaded it manually.
On any case, I think that, in the database table, rather than a type 'collection' we must have a column 'collection' that can be null for individually uploaded ones and the URL of the members of a collection uploaded (we 'expand' the collection into individual badges, on per record). This let us to do further queries for new content of each collection (the URL will get repeated, we just query for different ones). If something more than the collection URL is needed, we can store them in a new table and link back by ID (can be autogenerated or the URL itself since is unique). This allow us too to, at the listing editor, display the badges inside a collection grouped someway, letting user to remove the collection altogether or hide individual items, or requesting to refresh the collection (in case it did changes outside and our recurrent tasks still didn't catch).

--------------

I can visualize that request required badges by solutions can be too much, but at the same time this exposes the same problem of recording them by solution (too much, with duplications). We may need to come back to job title for required ones? Is it possible to use solution subcategories for this purpose?

**Comment by joshdanielson** [comment]
@iagosrl I maybe didn’t explain myself well enough. I meant that we manually approve the professional to provide a solution when credentials/ background check are required, NOT the badges themselves. So if you provide massage therapy, you have to upload a required credential (same process as now but perhaps different wording as it’s not based on location). Then, staff review what’s uploaded, issue a badge(s) and set the approved flag in userlistingsolution to 1. We’d set up a process to review expiring badges and this would go back to 0 if not renewed.

Does that make more sense?

**Comment by IagoSRL** [comment]
That makes sense, didn't get it correctly (now I see that says UserListingSolution, d'Oh!).

About naming, was thinking similar, if should be UserSolution for simplicity but really the relationship is user--listing--solution so the original one makes sense (and the records there will ever be related explicitly with a listing ID, right?)

Did you decided about to relate the badges with solutions or with listings?

**Comment by joshdanielson** [comment]
@IagoSRL after thinking about it more...I'm wondering how much extra work it would be to essentially add extra mapping tables for listing and solution. So hear me out...it would look something like this:
- UserSolution (this will be populated by users adding solutions from their listing but allow us to "reuse" solution data/mappings if users want to list solutions across multiple listings. Staff will still approve these manually for those solutions flagged as needing credentials or background checks)
- UserOffering ("providerpackage")
- UserSolutionOffering (mapping table...instead of mapping them to a listing, we map them to a solution.)
- UserBadge (where badges and collections are stored...users can upload and staff also adds to this based on licenses uploaded or background checks done)
- UserSolutionBadge (Users and/or staff attach badges to solutions)
- UserListingSolution (we will use this to pull solutions and their related offerings and badges)

Thoughts? I don't think anything (new) that we've done would prevent this but it may need some clarifications.

**Comment by IagoSRL** [comment]
I was thinking some similar regarding to UserSolution and badges on previous comment, but though if solutions were linked just to one job title and badges to one solution (except user-wide ones), then a mapping table makes no sense; if that's not the case, then we can go that way.
But there are hidden complexities, both for us implementing and the user. Major one is related with the feeling that, if I manage something at listing editor, it's scoped to *this* listing title, not all of them (when is shared).
Thinking: some editable things are "easier" to understand they are shared (because of the subject or position in the visual layout), like the profile photo, website, name (and still, there is margin to someone thinking if that should be specific per listing); easier to confuse are offerings in my opinions:
- If I remove this offering here, it gets removed from my other listing (where exists because the same solution is there)?
- If solution related things appear at both listings, Would the user think if makes more sense to have just one listing? (easier to manage and reason about it).

For me, a couple of requirements should be crystal clear before we continue discussing:
- Two solutions must not partially overlap each other. From user perspective, must be no-brainer to choose a solution where to add a specific offering, or a badge; don't force them to duplicate items too.
- How professionals will use the possibility to add more than one listing?
  - a) Doing different kind of work (offering different solutions)
  - b) Related work (some shared solutions): is a matter of marketing to have two or three listings sharing services offered (scope, targeting, campaigns)

For the first point I'm already giving my opinion about how I think about the 'solutions' idea, just let me know if is the opposite (overlapping solutions will exists). On that case, we need to study how much that makes user experience worse/complicated (both, professionals, gig posting clients, marketplace search clients).

**Comment by joshdanielson** [comment]
@IagoSRL good points. I think there may always be some complexity so we should just aim to reduce it with the goal of the underlying architecture being flexible enough to shift if we see a need for it in the future as we can't fully predict all use cases.

We tried to not make overlapping solutions but I think the Wedding category is the main, if not only, source of overlap now (look at the [SearchSubCategorySolution tab](https://docs.google.com/spreadsheets/d/1gDJPsJaDB66GIjhOYFeCP_Gnbw_hdgMQLNOBy7LgN0M/edit?usp=sharing) and scroll to the right). The thought was that just because you do photography, doesn't mean you do wedding photography (same for event planning, catering, etc.) but that Wedding is a popular category that we'd want it by itself. Really, "Wedding" is more of a specialization than a category but keeping it a category could help in search eventually. Always something. Otherwise, there isn't much overlap. We could eliminate all those solutions and the wedding category and then, when we get to it, create a category based on the professional having "Wedding" in their specialization, for example.

If you think that's not a good enough reason to keep wedding and the overlapping solutions (with a fix someday coming with specializations), then we could just remove them.

One thing to remember is that we plan to control what solutions are offered (not ever user generated) so that can alleviate some of the complexity.

**Comment by IagoSRL** [comment]
Sorry the delay on an answer, I have my doubts, even shifted to some parallel (but annoying) tasks and come back to think on this.

I thinks we should prevent overlapping, so Wedding specifics will get out now. Makes sense as a specialization. We can research (later, in an iteration of this mechanism)) how to promote it better in the search of solution, first thinking:
- making specializations (or someones we tag) to be part of the keywords for the search, so users can search by 'wedding' and get back a list with solutions that include that as specialization (like 'photography' --yep, wedding is not part of the results text)
- another iteration, make possible that we found relevant combinations os solution->specialization: searching by 'wedding' shows results as 'photography -> wedding' and others (can visually look as 'wedding (photography)' or any other we research makes sense).
- another idea is to make some kind of aliases for solutions, like searching by 'wedding' and displaying as result 'Wedding photography' (we manually put that text) by internally just filters by photography, and enhanced, by 'wedding' specialization. This looks more "hacky" and worse maintenance but allowing better language. Anyway, for later if chosen.

----------

About some complexities, some details I think we can go from simple to complex without break too much, thinking in the mappings. When we have a userListingID in a table like UserSolution (actually, UserListingSolution, but could get renamed now to reduce problems on changing this later), for the exposed REST API we just get all the solutions that apply to a listing (so will filter by that field); if we move to a mapping table, we migrate the data to the new table with userListingID--solutionID, that initially is one record for each, removing the userListingID from UserSolution and changing the queries code; for the results from the REST API, it's the same (done well).

Anyway, on this specific case I think we don't need to share solutions between listings if solutions don't overlap AND the usage of multiple listings by users is to have different kind of work on each one. So this last question, still unanswered, it's important. I think we can start by thinking that any listing is different but researching with our users if they pretend to create (or actually have created) listings pretty similar just to give them some marketing 'touch', like targeting a specialization, kind of clients; following the 'wedding' thing, can happens that a photographer has a general listing calling himself 'photographer' (listing title) and another 'wedding photographer'; at this case it will have shared solutions at both by sure, but different specializations chosen for each and different offerings (because is not the same) even if they match in the solution. We need data here, but thinking on this example It seems to me that is better to not share that things between listings (no mapping tables), but again, real data and use cases from users matters.

**Comment by joshdanielson** [comment]
@IagoSRL agree with all points and to remove wedding as there are better ways to accomplish that as you mentioned.

In terms of the duplication of solutions for listings, yes, I agree on waiting for more use cases. I guess the issue at hand now is that we want user to be able to manually add a solution to their listing (not predetermined) so in theory could overlap multiple listings.  

As long as we've organized the tables as UserSolution and UserListingSolution, we can respond to those use cases better should we need to.

**Comment by IagoSRL** [comment]
OK, good! So we go with offerings and solutions attached just to one listing (not shared).

I kept thinking that users can only add custom specializations but not solutions. And I prefer it that way. I think we can add some kind of 'request solution' if user wants to add something custom, we get an email about what the user proposed while we communicate that we will review the proposal and answer back, or something like that, but without add it to the table. Makes sense?

About tables UserSolution and UserListingSolution, only one of that names must exists (right now, the latest). We can rename it as UserSolution just in case we add a mapping table in the future (so is easier) but keeping it now with a UserListingID field. That's the question?

**Comment by joshdanielson** [comment]
@iagosrl okay, maybe not quite there...

Agree with the request a solution and do it in a help ticket, nothing user generated but yes, allow it in specialization but only show ones that are approved.

I still think we should have usersolution, useroffering, usersolutionoffering and then userlistingsolution and avoid mapping in the first two even if the latter two don’t have one-to-many relationships (yet) as it makes it much clearer the relationships and would allow us to create the abilities in the future without having to redo the architecture. Is that much more work? Again, I’m not an expert but it feels like it’s better to start predicting some complexity if there isn’t much difference.

**Comment by IagoSRL** [comment]
As we experienced in the past, I think we should prevent complexity if we are no going to make use of it. Because it ever ends to be more work, code paths and edge cases leading to crashes or wrong app state if not managed, even if not useful for the subset being used.

But if you want to go that way, then go with it; don't want to keep taking time discussing this if the conclusion will at the end being the same, anyway. Just explain in a comment the final tables, with fields as a bullet list and which one makes relations, because are now getting confusing for me (don't understand the comment about we don't have one-to-many relationship but we have tables that suggest we have that).

**Comment by joshdanielson** [comment]
@IagoSRL okay, I hope this covers everything but please amend as you see fit. 
- [ ] Rename UserListingSolution to UserSolution (in case we create mappings later). Should we include an auto-incremental UserSolutionID to enable a primary key not related to listings in the future? I proposed one in UserBadge but not sure your thoughts on it.
- [ ] Add SolutionID to providerpackage
- [ ] Create [UserBadge table](https://docs.google.com/spreadsheets/d/1D1AnVWKXOG2w6WkeGPTSL0-51Xn_NugbrPlcpvqu_P8/edit?usp=sharing) that includes SolutionID field (can be -1 for generic badges). 

Let me know if something isn't clear.

Please assign me anything you'd like me to take on.

**Comment by IagoSRL** [comment]
@joshdanielson I agree with that three points, and to have an auto incremental ID at both tables

**Comment by IagoSRL** [comment]
@joshdanielson Please do any organization of the solutions data and related so it matches the expectations we commented last days (like no overlapping solutions)

**Comment by IagoSRL** [comment]
@joshdanielson I updated this branch with the content from master. I had to update a conflict at listingEditor cause of the fix with the layout we did past week (the text displayed on empty content that appeared narrowed at the right); I merged this by using the 'new' URL for the add button (credentials-add) while keeping the previous one for the edit button (licensesCertifications) because there was not edit button no edit activity here still. And the text for empty data gets added, just review if makes sense for this credentials/badges thing.
I will ask you as task to review and fix the layout, visualization of this bagdes components: they appear or too big (images) or with a broken markup or layout (latest testing elements).
I will do the table changes at the other epic branch and merge to master, then here.

**Comment by IagoSRL** [comment]
I make the database changes, applied to *dev* from the scripts I created/updated (at branch is796).
Details: rather than using -1 for SolutionID at UserBadge, I make the field nullable, to no need to add the special ID to the Solutions table (and works the same, just checking for null value understood that badge applies to the whole user profile); added languageID and countryID columns to make the FK with solution table.

**Comment by joshdanielson** [comment]
@IagoSRL tested both a single badge and collection. The single badge worked great. The collection display is a bit off and only the first badge appears well:
![screen shot 2018-04-03 at 2 29 49 pm](https://user-images.githubusercontent.com/1202838/38277152-40926f00-374c-11e8-8590-d07f740ead64.png)

**Comment by IagoSRL** [comment]
*Continuing highly related https://github.com/loconomics/loconomics/issues/747#issuecomment-378532535:*
I used [this collections URL](https://api.badgr.io/public/collections/fcfad6d9e51bd635c9c88248587dc035?v=2_0 and) and it worked fine displaying all the badges inside.
The collection you are trying seems to return some unexpected format for the 'evidence' data (not null and not an array), throwing an error, see the console at your screenshot, it says that code `(assertion.evidence || []).filter` *is not a function*. In case no evidence (is null or undefined) I fallback to an empty array (that has a 'filter' function), but for some reason fails here, like if 'evidence' property exists but contains something that is not an array, so built-in 'filter' function does not exist on it.
**Please send me the URL you are trying** to review it; and try with the previous URL I posted (that is yours actually) to see that it works with that.

**Comment by joshdanielson** [comment]
@IagoSRL badges look great! The add/edit buttons aren't aligned now, though (I don't think the pull-right class works with container-fluid):
![screencapture-localhost-8811-2018-04-10-08_06_09](https://user-images.githubusercontent.com/1202838/38565518-5ff23afc-3c96-11e8-9448-4cdfa5624d19.png)

**Comment by IagoSRL** [comment]
I see now. Actually, the problem is that SectionHeading has an auto width, so shrinks to just fit what needed (depending on internal elements width), and container-fluid too. But col-*-12 classes have `width: 100%;` so that make it. I will put that full width directly to the SectionHeading, I think makes sense since is expected that on *any* case is used must take the full width; if shouldn't take all the width, then the container where it's must be the one that reduces the width (with columns for example). Do you think this could case a problem at other use cases of this class?

**Comment by IagoSRL** [comment]
I pushed that change, so you can give it a try

**Comment by joshdanielson** [comment]
@IagoSRL looks good. I poked around the rest of the app and didn't see any issues.

**Comment by IagoSRL** [comment]
@joshdanielson In the branch I replaced the user licences/certifications with the badges at the public listing (in addition to listingEditor); please give it a try since I think is the *last* thing in order to close this.

**Comment by IagoSRL** [comment]
Well, I commented here on the epic rather than the relevant issues #748 #749, check the others (assigned to @joshdanielson) so we close the epic only after close all of them.

**Comment by IagoSRL** [comment]
Please @joshdanielson review the URL /admin at this branch, with an admin editor of user badges.

The list could be a table-like view with record data rather than current usage of the full badge view component but it does the work.
The user input can get replaced with a service proffesionals autocomplete.
Impl. details: some activity code should go in admin components (only there is one now, the badge-admin-editor borrowing the badge-editor).

Let me know if does what is expected/required (knowing that previous drawbacks)

**Comment by joshdanielson** [comment]
@iagosrl tested and works great on the branch. More than good enough for now.

**Comment by IagoSRL** [comment]
Good! Next steps on this Epic?

**Comment by joshdanielson** [comment]
@IagoSRL I'll finish #770 by adding instructions for the admin tool. #749 will be ongoing so may be okay to close for now and create issues for specific badges as they arise (done in Badgr so no further development required).

**Comment by joshdanielson** [comment]
@IagoSRL sorry!! Not sure if we should add the solutionID to the /admin-user-badges form now or wait?

**Comment by IagoSRL** [comment]
I enabled the solutionID at the REST API, and put some commented code for it at both badge-editor and badge-admin-editor. Will enable when solutions are enabled, placing there a solution-autocomplete.

**Comment by joshdanielson** [comment]
@iagosrl great!

**Comment by IagoSRL** [comment]
Released (webapp v1.10)
